presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/BingTileUtils.java

```diff
@@ -168,4 +186,163 @@ private static int axisToCoordinates(double axis, long mapSize)
+                BingTile.fromCoordinates(1, 0, 1),
+                BingTile.fromCoordinates(1, 1, 1)
+        ).forEach(addIntersecting);
+
+        List<BingTile> outputTiles = new ArrayList<>();
+        while (!stack.isEmpty()) {
+            TilingEntry entry = stack.pop();
+            if (entry.tile.getZoomLevel() == maxZoom || ogcGeometry.contains(entry.polygon)) {
+                outputTiles.add(entry.tile);
+            }
+            else {
+                entry.tile.findChildren().forEach(addIntersecting);
+                checkCondition(
+                        outputTiles.size() + stack.size() <= MAX_COVERING_COUNT,
+                        "The zoom level is too high or the geometry is too large to compute a set " +
+                                "of covering Bing tiles. Please use a lower zoom level, or tile only a section " +
+                                "of the geometry.");
+            }
+        }
+        return outputTiles;
+    }
+
+    private static Optional<List<BingTile>> handleTrivialCases(Envelope envelope, int zoom)
+    {
+        checkZoomLevel(zoom);
+
+        if (envelope.isEmpty()) {
+            return Optional.of(ImmutableList.of());
+        }
+        checkLatitude(envelope.getYMin(), LATITUDE_SPAN_OUT_OF_RANGE);
+        checkLatitude(envelope.getYMax(), LATITUDE_SPAN_OUT_OF_RANGE);
+        checkLongitude(envelope.getXMin(), LONGITUDE_SPAN_OUT_OF_RANGE);
+        checkLongitude(envelope.getXMax(), LONGITUDE_SPAN_OUT_OF_RANGE);
+
+        if (zoom == 0) {
+            return Optional.of(ImmutableList.of(BingTile.fromCoordinates(0, 0, 0)));
+        }
+        if (envelope.getXMax() == envelope.getXMin() && envelope.getYMax() == envelope.getYMin()) {
+            return Optional.of(ImmutableList.of(latitudeLongitudeToTile(envelope.getYMax(), envelope.getXMax(), zoom)));
+        }
+
+        return Optional.empty();
+    }
+
+    /*
+     * BingTiles don't contain their eastern/southern edges, so that each point lies
+     * on a unique tile.  However, the easternmost and southernmost tiles must contain
+     * their eastern and southern bounds (respectively), because they are the only
+     * tiles that can.
+     */
+    private static boolean satisfiesTileEdgeCondition(Envelope query, TilingEntry entry)
+    {
+        BingTile tile = entry.tile;
+        int maxXY = (1 << tile.getZoomLevel()) - 1;
+        if (tile.getY() < maxXY && query.getYMax() == entry.envelope.getYMin()) {
+            return false;
+        }
+        if (tile.getX() < maxXY && query.getXMin() == entry.envelope.getXMax()) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Find a minimal set of BingTiles (at different zooms), covering the geometry.
+     * If a larger tile fits within the geometry, do not split it into smaller
+     * tiles.  Do not split a tile past maxZoom.
+     */
+    public static List<BingTile> findDissolvedTileCovering(OGCGeometry ogcGeometry, int maxZoom)
```

Comment:
nit: please move it before the `findRawTileCovering `, as `findDissolvedTileCovering ` uses `findDissolvedTileCovering `